src/StreamJsonRpc.Tests/JsonRpcProxyGenerationTests.cs

```diff
@@ -0,0 +1,343 @@
+    }
+
+    public interface IServer
+    {
+        event EventHandler ItHappened;
+
+        event EventHandler<CustomEventArgs> TreeGrown;
+
+        Task<string> SayHiAsync();
+
+        Task<string> SayHiAsync(string name);
+
+        Task<int> AddAsync(int a, int b);
+
+        Task IncrementAsync();
+
+        Task Dispose();
+
+        Task HeavyWorkAsync(CancellationToken cancellationToken);
+
+        Task<int> HeavyWorkAsync(int param1, CancellationToken cancellationToken);
+
+        [JsonRpcMethod("AnotherName")]
+        Task<string> ARoseByAsync(string name);
+    }
+
+    public interface IServer2
+    {
+        Task<int> MultiplyAsync(int a, int b);
+    }
+
+    public interface IServerWithNonTaskReturnTypes
+    {
+        int Add(int a, int b);
+    }
+
+    public interface IServerWithUnsupportedEventTypes
+    {
+        event Action MyActionEvent;
+    }
+
+    public interface IServerWithProperties
+    {
+        int Foo { get; set; }
+    }
+
+    public interface IServerWithGenericMethod
+    {
+        Task AddAsync<T>(T a, T b);
+    }
+
+    internal interface IServerInternal
+    {
+        Task AddAsync(int a, int b);
+    }
+
+    [Fact]
+    public void ProxyTypeIsReused()
+    {
+        var streams = FullDuplexStream.CreateStreams();
+        var clientRpc = JsonRpc.Attach<IServer>(streams.Item1);
+        Assert.IsType(this.clientRpc.GetType(), clientRpc);
+    }
+
+    [Fact]
+    public async Task CallMethod_String_String()
+    {
+        Assert.Equal("Hi, Andrew!", await this.clientRpc.SayHiAsync("Andrew"));
+    }
+
+    [Fact]
+    public async Task CallMethod_void_String()
+    {
+        Assert.Equal("Hi!", await this.clientRpc.SayHiAsync());
+    }
+
+    [Fact]
+    public async Task CallMethod_IntInt_Int()
+    {
+        Assert.Equal(3, await this.clientRpc.AddAsync(1, 2));
+    }
+
+    [Fact]
+    public async Task CallMethod_void_void()
+    {
+        await this.clientRpc.IncrementAsync();
+        Assert.Equal(1, this.server.Counter);
+    }
+
+    [Fact]
+    public void ImplementsIDisposable()
+    {
+        var disposableClient = (IDisposable)this.clientRpc;
+        disposableClient.Dispose();
+        Assert.True(this.clientStream.IsDisposed);
+    }
+
+    [Fact]
+    public async Task DisposeCollision()
+    {
+        // We're calling IServer.Dispose -- NOT IDisposable.Dispose here.
+        // Verify that it invokes the server method rather than disposing of the client proxy.
+        await this.clientRpc.Dispose();
+        Assert.Equal(-1, this.server.Counter);
+        Assert.False(this.clientStream.IsDisposed);
+    }
+
+    [Fact]
+    [Trait("NegativeTest", "")]
+    public void NonTaskReturningMethod()
+    {
+        var streams = FullDuplexStream.CreateStreams();
+        var exception = Assert.Throws<NotSupportedException>(() => JsonRpc.Attach<IServerWithNonTaskReturnTypes>(streams.Item1));
+        this.Logger.WriteLine(exception.Message);
```

Comment:
Because the exception messages themselves are usually formatted, and just testing the exception *type* doesn't really test that the message was formatted properly. Writing automated tests to validate the message is tough to get to pass/fail consistently when running on non-ENU machines where translated messages can often make a message test fail.
So, I took the easy way out and the test writes the message out so I can just eyeball it to make sure it's doing it right.